Add session arguments

crleblanc opened the pull request:
This pull request adds missing options for the Session object (eg: pyFiles, files, archives, etc) as defined in the Livy documentation (https://livy.incubator.apache.org/docs/latest/rest-api.html).  My project requires the pyFiles argument to reference additional Python modules available on S3.  I think this enhancement would be of use to many people in the Python/Livy community.

This is based on #47 and #34, rebased on the latest commit of the master branch and updated to resolve issues with mypy and black.  There is no additional logic, just passing arguments to the Livy request.  I have tested the py_files option on a Spark EMR v2.4.0 running Livy v0.5.0 and it worked as expected.

Thanks to Parisni for his work on the original PR.

acroz commented:
Thanks @crleblanc for looking into this and @parisni for the original PR.

This will be a great extension to the capabilities of the library. Before merging, however, I'd like to have unit tests for the `LivyClient` that check that the correct payload is passed when these values are passed. Could you look at extending `test_create_session` in `test_client` to pass representative values for these options and check they get sent in the request in the correct format?

acroz commented:
The main reason I didn't add these in the first version is that I didn't fully understand the meaning and/or suitable values to be passed. I know the options added here have been documented using Livy's REST documentation, but the following options I find are vaguely commented and could use improvement:

```
:param jars: jars to be used in this session
:param py_files: Python files to be used in this session
:param files: files to be used in this session
:param archives: Archives to be used in this session
```

Can we make these descriptions more clear in some way? e.g. What do 'files' or 'archives' actually do?

On a similar note, I note that `driver_memory` and `executor_memory` are formatted as strings, but the Livy REST API docs give no information about how these strings are formatted. Can we make this clearer in our docstrings?

acroz commented:
Final note for now: I've just merged another PR which will allow you to run tests locally with `tox` - you may wish to rebase if doing any test changes.

crleblanc commented:
Hi @acroz,

Thanks for having a look at this.  I'd be happy to make a unit test to check the payload, should be pretty straightforward.

I also found the Livy documentation to be very sparse for these options.  I think I'll have to experiment with these options and look at the Livy source code to get a better idea.  I'll update the docstrings with more helpful information.

I'll also rebase to the latest version.

acroz commented:
Thanks, though if it turns out to be a pain to determine the meaning of those options, I'm happy to accept the REST API descriptions for the time being and leave an open issue for someone with experience of these options to clarify later.

Maybe @parisni can help with this since he wanted the `py_files` option in the first place?

crleblanc commented:
I've rebased to the latest version of master and added the additional arguments to the test_client.py test and it still passes.

I also researched these Livy session options and examined the Livy logs to see what it's doing.  I think the docstrings should be much more informative now, especially for options like 'py_files' and 'driver_memory'.  I've also copied some relevant information from the Spark docs and Livy Java/Spark API docs.

Hopefully this is in good shape now, let me know if there's anything else I can do.  Feel free to make any modifications you see fit.

Cheers,
Chris

acroz commented:
This is now released to PyPI in 0.5.0 - thanks for the contribution!
